README.md
<div id="top"></div>

<!-- PROJECT SHIELDS -->
<!--
*** https://www.markdownguide.org/basic-syntax/#reference-style-links
-->

<br />
<div align="center">
  <a href="https://github.com/xlab-ub/xlab-web">
    <img src="img/doc/website.svg" alt="Logo" width="120" height="120">
  </a>

<h3 align="center">X-Lab's Website</h3>

</div>



<details>
  <summary>Table of Contents</summary>
  <ol>
    <li>
      <a href="#about-the-project">About The Project</a>
    </li>
    <li>
      <a href="#getting-started">Getting Started</a>
      <ul>
        <li><a href="#useful-links">Useful Links</a></li>
      </ul>
    </li>
    <li><a href="#left-menu">Left Menu</a></li>
    <li>
      <a href="#sections">Sections</a>
      <ul>
        <li><a href="#home-section">Home Section</a></li>
        <li><a href="#news-section">News Section</a></li>
        <li><a href="#research-section">Research Section</a></li>
        <li><a href="#publication-section">Publication Section</a></li>
        <li><a href="#recognitions-section">Recognitions Section</a></li>
        <li><a href="#services-section">Services Section</a></li>
        <li><a href="#teaching-section">Teaching Section</a></li>
        <li><a href="#team-section">Team Section</a></li>
        <li><a href="#positions-section">Positions Section</a></li>
        <li><a href="#contact-section">Contact Section</a></li>
      </ul>
    </li>
    <li>
      <a href="#pages">Pages</a>
      <ul>
        <li><a href="#patents-page">Patents Page</a></li>
        <li><a href="#journal-page">Journal Page</a></li>
        <li><a href="#conferences-page">Conferences Page</a></li>
        <li><a href="#workshops-page">Workshops Page</a></li>
        <li><a href="#invited-papers-page">Invited Papers Page</a></li>
        <li><a href="#books-and-chapters-page">Books and Chapters Page</a></li>
      </ul>
    </li>
    <li><a href="#license">License</a></li>
    <li><a href="#contact">Contact</a></li>
  </ol>
</details>


<br />


## About The Project

<p align="right">(<a href="#top">back to top</a>)</p>



## Getting Started

This document describes how you can use the website. This project includes three parts: **left menu**, **sections**, and **pages**.

### Useful Links

- Markdown Guide
  - https://www.markdownguide.org/basic-syntax/#reference-style-links
- Add classes, identifiers and attributes to your markdown
  - https://github.com/arve0/markdown-it-attrs
- Add block-level custom containers and Div tags to your markdown
  - https://github.com/markdown-it/markdown-it-container or https://github.com/kickscondor/markdown-it-div
- Add Span tags to your markdown
  - https://github.com/mb21/markdown-it-bracketed-spans
- Email Encoders
  - https://ctrlq.org/encode/
  - https://www.knechtology.com/stop-spam/email_encoder.html
- Use Bootstrap for styling your contents
  - https://getbootstrap.com/docs/
- Use Font Awesome for adding icons to your contents
  - https://fontawesome.com/search

<p align="right">(<a href="#top">back to top</a>)</p>

## Left Menu
You can edit `index.html` to update the left menu. To do this, please edit the contents inside `<header id="site_header" class="header mobile-menu-hide"> ... </header>` tag.

<p align="right">(<a href="#top">back to top</a>)</p>


## Sections

Sections are the main contents of the website. To edit each section, you must edit the corresponding markdown file in the `sections` folder.

_NOTE: You can use a combination of `HTML` and `MARKDOWN` in the markdown files._



### Home Section

You must edit the `sections/home.md` file to update the logo, title, and contents of the home section.

### News Section

You must edit the `sections/news.md` file to update the contents of the news section.

Follow up these steps to add news to your website:

1. Put this template below `<!--Put your news below this line -->` line in the `sections/news.md` file:
```
<div class="timeline timeline-second-style clearfix">
    <div class="timeline-item clearfix pb-3">
        <div class="left-part">
            <h5 class="item-period"><b>{{DATE}}</b></h5>
            <span class="item-company"><span class="badge bg-success"><i class="fa-regular fa-newspaper"></i> <b>{{TYPE}}</b></span></span>
        </div>
        <div class="divider"></div>
        <div class="right-part">
            <h4 class="item-title"><b>{{TITLE}}</b></h4>
            <p><b>{{DETAIL}}</b></p>
        </div>
    </div>
</div>
```
2. Replace `{{DATE}}`, `{{TYPE}}`, `{{TITLE}}`, and `{{DETAIL}}` variables with your contents.
3. To change the color of the `{{TYPE}}` variable, please replace `<span class="badge bg-success"> ... </span>` with an another color from https://getbootstrap.com/docs/5.2/components/badge/#background-colors
4. To change the icon of the `{{TYPE}}` variable, please replace `<i class="fa-regular fa-newspaper"></i>` with a new icon from https://fontawesome.com/search


_NOTE: The `{{DETAIL}}` variable is optional. Therefore, if your news hasn't `{{DETAIL}}` variable, you must remove `<p>{{DETAIL}}</p>`._

### Research Section

You must edit the `sections/research.md` file to update the contents of the research section.

### Publication Section

This section includes some sub-sections. To create a new sub-section, you must put the following template below `<!--Put your item below this line -->` line in the `sections/publication.md` file and change the variables (`{{...}}`) based on your needs:

```
<div class="col mt-3">
    <div class="card" style="width: 13rem;">
        <img src="./img/{{ICON_FILE}}.svg" class="card-img-top" alt="{{ALT_TEXT}}" style="padding: 1rem;" title="{{TITLE_TEXT}}">
        <div class="card-body text-center">
            <a href="pages/{{PAGE_NAME}}.html" class="btn btn-primary ajax-page-load" style="font-size: .85em;">{{BUTTON_TEXT}}</a>
        </div>
    </div>
</div>
```

_NOTE: Edit variables (`{{...}}`) based on your needs._

_NOTE2: To create a new page, create a new `HTML file (*.html)` in `pages` folder and replace its name with `{{PAGE_NAME}}` variable._

### Recognitions Section

You must edit the `sections/recognitions.md` file to update the contents of the recognitions section.

To add a new accordion item, please put the following template below `<!-- Put the accordion item below this line -->` line in the `sections/recognitions.md` file:

```
<div class="accordion-item">
  <h2 class="accordion-header" id="{{ACCORDION_ID}}">
    <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#{{CONTAINER_ID}}" aria-expanded="true" aria-controls="{{CONTAINER_ID}}">
      {{TITLE}}
    </button>
  </h2>
  <div id="{{CONTAINER_ID}}" class="accordion-collapse collapse" aria-labelledby="{{ACCORDION_ID}}" data-bs-parent="#recognitionsAccordion">
    <div class="accordion-body">

{{CONTENT}}
    </div>
  </div>
</div>
```

_NOTE: Edit variables (`{{...}}`) based on your needs._

_NOTE2: You must replace your content with `{{CONTENT}}` variable. The content can be a `MARKDOWN` or `HTML`._

### Services Section

You must edit the `sections/services.md` file to update the contents of the services section.

To add a new accordion item, please put the following template below `<!-- Put the accordion item below this line -->` line in the `sections/services.md` file:

```
<div class="accordion-item">
  <h2 class="accordion-header" id="{{ACCORDION_ID}}">
    <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#{{CONTAINER_ID}}" aria-expanded="true" aria-controls="{{CONTAINER_ID}}">
      {{TITLE}}
    </button>
  </h2>
  <div id="{{CONTAINER_ID}}" class="accordion-collapse collapse" aria-labelledby="{{ACCORDION_ID}}" data-bs-parent="#recognitionsAccordion">
    <div class="accordion-body">

{{CONTENT}}
    </div>
  </div>
</div>
```

_NOTE: Edit variables (`{{...}}`) based on your needs._

_NOTE2: You must replace your content with `{{CONTENT}}` variable. The content can be a `MARKDOWN` or `HTML`._

### Teaching Section

You must edit the `sections/teaching.md` to update the contents of the teaching section.

To add a new course, please put the following template below `<!-- Put your course below this line -->` line in the `sections/teaching.md` file:

```
<tr>
  <th scope="row">{{COURSE_NUMBER}}</th>
  <td>{{COURSE_NAME}}</td>
  <td>{{SEMESTERS}}</td>
  <td>{{LOCATION}}</td>
</tr>
```

_NOTE: Edit variables (`{{...}}`) based on your needs._

### Team Section

You must edit the `sections/team.md` to update the contents of the teaching section.

To create a new block in the team page, please use the following template in the `sections/team.md` file:

```
<div class="text-center"><div class="block-title"><h3>{{BLOCK_TITLE}}</h3></div></div>
  <div class="row row-cols-auto g-4 justify-content-center">
    <!-- Put the new member item below this line -->

  </div>
</div>
```

To add a new team member, please put the following template below `<!-- Put the new member item below this line -->` line in the top template or `inside the blocks` in the `sections/team.md` file: 

```
<div class="col">
  <div class="card-team">
    <img src="./img/team/{{PERSON_DISPLAY_PHOTO_NAME}}" />
    <div class="info">
      <h1>{{FULL_NAME}}</h1>
      <p>{{MEMBER_TITLE}}</p><a href="{{PERSONAL_URL}}" target="_blank">{{URL_TEXT}} <i class="fa-solid fa-caret-right"></i></a> <span data-bs-toggle="tooltip" data-bs-html="true" data-bs-custom-class="custom-info-tooltip" title="<p class='text-left text-warning'>{{ENCODED_EMAIL_ADDRESS}}</p><hr class='mt-0'><p class='text-left'>{{INFO}}</em></p>" class="info-button d-inline-block float-end"><i class="fa-solid fa-circle-info fa-lg fa-beat-fade" style="--fa-beat-fade-opacity: 0.67; --fa-beat-fade-scale: 1.075;"></i></span>
    </div>
  </div>
</div>
```

_NOTE1: `{{BLOCK_TITLE}}` variable is title of a block, for example: Ph.D. Students, Master Students, etc._

_NOTE2: `{{PERSON_DISPLAY_PHOTO_NAME}}` must be equivalent to **full name** of a person's photo which is uploaded in `img/team/` folder._

_NOTE3: If a person hasn't any of the `{{PERSONAL_URL}}` or `{{INFO}}` variables, please remove them from the template accordingly._

_NOTE4: You should encode a person's email by https://ctrlq.org/encode/ and replace it with `{{ENCODED_EMAIL_ADDRESS}}` variable._

### Positions Section

You must edit the `sections/positions.md` file to update the contents of the positions section.

### Contact Section

You must edit the `sections/contact.md` file to update the contents of the contact section.

_NOTE: Please encode your email by one of the `Email Encoders` to prevent spam._


<p align="right">(<a href="#top">back to top</a>)</p>


## Pages
Pages contain some contents, which can be loaded by ajax calls, for example, when a user clicks on a button. Please check the `Publication` section for a live example.

Pages must be an `HTML` file.

### Patents Page
To add a new item to the patents page, you can put the item as a `<li>...</li>` tag below `<!-- Add a new item below this line -->` line in the `pages/patents.html` file.

### Journal Page
You must edit the `pages/journal.html` file to update the contents of the journal page.

To add a new paper, please put the following template below `<!-- Put your paper below this line -->` line in the `pages/journal.html` file:

```
<div class="row stacked-list">
    <div class="col-sm-12 col-md-12 portfolio-block pt-0">
        <div class="certificate-item clearfix mb-0 radius-0">
            <div class="certi-logo radius-0 p-0 width-200">
                <a href="./img/publications/journals/{{PAPER_PHOTO}}" target="_blank"><img src="./img/publications/journals/{{PAPER_PHOTO}}" alt="logo"></a>
            </div>

            <div class="certi-content">
                <div class="certi-title">
                    <h4>{{PAPER_REFERENCE}}</h4>
                </div>
            </div>
        </div>
    </div>
</div>
```

_NOTE1: Edit variables (`{{...}}`) based on your needs._

_NOTE2: `{{PAPER_PHOTO}}` must be equivalent to **full name** of a paper's photo, which is uploaded in the`img/publications/journals/` folder._


### Conferences Page
To add a new item to the conferences page, you can put the item as a `<li>...</li>` tag below `<!-- Add a new item below this line -->` line in the `pages/conferences.html` file.

### Workshops Page
To add a new item to the workshops page, you can put the item as a `<li>...</li>` tag below `<!-- Add a new item below this line -->` line in the `pages/workshops.html` file.

### Invited Papers Page
To add a new item to the invited papers page, you can put the item as a `<li>...</li>` tag below `<!-- Add a new item below this line -->` line in the `pages/invitedpapers.html` file.

### Books and Chapters page
To add a new item to the books and chapters page, you can put the item as a `<li>...</li>` tag below `<!-- Add a new item below this line -->` line in the `pages/bookchapters.html` file.


<p align="right">(<a href="#top">back to top</a>)</p>

## License

Distributed under the Apache License 2.0. See the `LICENSE` file for more information.

<p align="right">(<a href="#top">back to top</a>)</p>


## Contact


<p align="right">(<a href="#top">back to top</a>)</p>



<!-- MARKDOWN LINKS & IMAGES -->
<!-- https://www.markdownguide.org/basic-syntax/#reference-style-links -->
[product-screenshot]: img/doc/screenshot.jpeg


sections/news.md
<div class="block-title"><h3>Be informed!</h3></div>

<!-- <ul class="news-list">
    <li class="item">
        <div>
            <p class="date">2020/4/15</p>
            <p class="category"><span class="badge bg-success"><i class="fa-solid fa-rss"></i> News</span></p>
            <p class="title"><span class="badge bg-danger">New</span> This is an event! <span class="details">Test ... test test test.</span></p>
            </br>
        </div>
    </li>
    <li class="item">
        <div>
            <p class="date">2020/4/15</p>
            <p class="category"><span class="badge bg-info text-dark"><i class="fa-solid fa-circle-info"></i> Information</span></p>
            <p class="title">This is an announcement!</p>
        </div>
    </li>
    <li class="item">
        <div>
            <p class="date">2020/4/15</p>
            <p class="category"><span class="badge bg-warning text-dark"><i class="fa-regular fa-calendar-check"></i> Event</span></p>
            <p class="title">This is an event!</p>
        </div>
    </li>
</ul> -->



<!-- <div>
    <p class="date">2020/4/15</p>
    <p class="category"><span class="badge bg-success"><i class="fa-solid fa-rss"></i> News</span></p>
    <p class="title"><span class="badge bg-danger">New</span> This is an event! <span class="details">Akdj skdjf bksjdfkjdshfjhkdfdfgdfg dfg.dfg dfgdf.g dfg.dfg. dfgdfgdfg dfgd.</span></p>
    </br>
</div> -->

<!-- <span class="item-company"><span class="badge bg-primary"><i class="fa-solid fa-rss"></i> News</span></span>
<span class="item-company"><span class="badge bg-warning text-dark"><i class="fa-regular fa-calendar-check"></i> Event</span></span> -->

<div class="row">
    <div class="col-xs-12 col-sm-12 col-12">
        <!--Put your news below this line -->
        <div class="timeline timeline-second-style clearfix">
            <div class="timeline-item clearfix pb-3">
                <div class="left-part">
                    <h5 class="item-period">6/12/2022</h5>
                    <span class="item-company"><span class="badge bg-success"><i class="fa-regular fa-newspaper"></i> Publication</span></span>
                </div>
                <div class="divider"></div>
                <div class="right-part">
                    <p class="item-title">It is our pleasure to inform you that our paper: <a href="https://arxiv.org/abs/2204.01701" target="_blank">QuadraLib: A Performant Quadratic Neural Network Library for Architecture Optimization and Design Exploration</a> is recognized as one of the top 5 outstanding papers in the <a href="https://mlsys.org/Conferences/2022/Schedule?showEvent=2059" target="_blank">MLSys</a> conference.</p>
                </div>
            </div>
        </div>
         <div class="timeline timeline-second-style clearfix">
            <div class="timeline-item clearfix pb-3">
                <div class="left-part">
                    <h5 class="item-period">3/12/2022</h5>
                    <span class="item-company"><span class="badge bg-success"><i class="fa-regular fa-newspaper"></i> Award</span></span>
                </div>
                <div class="divider"></div>
                <div class="right-part">
                    <p class="item-title">It is our pleasure to inform you that our team won 1st place at the 2022 <a href="https://mlsys.org/Conferences/2022/Schedule?showEvent=2059" target="_blank">MLSys</a> conference.</p>
                </div>
            </div>
        </div>
    </div>
</div>

js/custom.js
function customScroller() {
    var windowWidth = $(window).width();
    if (windowWidth > 1024) {
        $('.animated-section, .single-page-content').each(function () {
            $(this).perfectScrollbar('update');
        });
    } else {
        $('.animated-section, .single-page-content').each(function () {
            $(this).perfectScrollbar('destroy');
        });
    }
}



document.addEventListener("DOMContentLoaded", function () {
    axios.all([axios.get("./sections/home.md"),
            axios.get("./sections/news.md"),
            axios.get("./sections/research.md"),
            axios.get("./sections/publication.md"),
            axios.get("./sections/recognitions.md"),
            axios.get("./sections/services.md"),
            axios.get("./sections/teaching.md"),
            axios.get("./sections/team.md"),
            axios.get("./sections/positions.md"),
            axios.get("./sections/contact.md")
        ])
        .then(axios.spread((zeroResponse, firstResponse, secondResponse, thirdResponse, forthResponse, fifthResponse, sixthResponse, seventhResponse, eighthResponse, ninthResponse) => {
            var md = new window.markdownit("commonmark", {html: true, breaks: true, linkify: true, typographer: true});
            var mdAttr = window.markdownItAttrs;
            md.use(mdAttr);


            var elem0 = document.querySelector('#home-container');
            var html0 = md.render(zeroResponse.data);
            elem0.innerHTML = html0;

            var elem1 = document.querySelector('#news-container');
            var html1 = md.render(firstResponse.data);
            elem1.innerHTML = html1;

            var elem2 = document.querySelector('#research-container');
            var html2 = md.render(secondResponse.data);
            elem2.innerHTML = html2;

            var elem3 = document.querySelector('#publication-container');
            var html3 = md.render(thirdResponse.data);
            elem3.innerHTML = html3;

            var elem4 = document.querySelector('#recognitions-container');
            var html4 = md.render(forthResponse.data);
            elem4.innerHTML = html4;

            var elem5 = document.querySelector('#services-container');
            var html5 = md.render(fifthResponse.data);
            elem5.innerHTML = html5;

            var elem6 = document.querySelector('#teaching-container');
            var html6 = md.render(sixthResponse.data);
            elem6.innerHTML = html6;

            var elem7 = document.querySelector('#team-container');
            var html7 = md.render(seventhResponse.data);
            elem7.innerHTML = html7;

            var elem8 = document.querySelector('#positions-container');
            var html8 = md.render(eighthResponse.data);
            elem8.innerHTML = html8;

            var elem9 = document.querySelector('#contact-container');
            var html9 = md.render(ninthResponse.data);
            elem9.innerHTML = html9;


            customScroller();

            const tooltipTriggerListAjax = document.querySelectorAll('[data-bs-toggle="tooltip"]');
            const tooltipListAjax = [...tooltipTriggerListAjax].map(tooltipTriggerEl => new bootstrap.Tooltip(tooltipTriggerEl));
        
          }))
        .catch(error => console.log(error));
}, false);

const tooltipTriggerList = document.querySelectorAll('[data-bs-toggle="tooltip"]');
const tooltipList = [...tooltipTriggerList].map(tooltipTriggerEl => new bootstrap.Tooltip(tooltipTriggerEl));


js/settings.js
$(document).ready(function () {
    "use strict";
    try {
        

        /* ==========================================================================
         #Carousel Popup For Portfolio Page
         ========================================================================== */
        $(".owl-carousel").owlCarousel({
            navigation: true,
            slideSpeed: 300,
            paginationSpeed: 400,
            singleItem: true,
            autoPlay: false
        });

        /* ==========================================================================
         #Text Rotator 
         ========================================================================== */
        $('#rotate').rotaterator({fadeSpeed: 800, pauseSpeed: 800});

        /* ==========================================================================
         #Orientation change event
         ========================================================================== */
        $(window).on('orientationchange', function (event) {
            window.location.href = window.location.href;
        });

        //Videos
        $(".content-scroller").fitVids();

    } catch (ex) {
    }
});

/* ==========================================================================
 #Progress Bar For Skills Page
 ========================================================================== */

progressBar(99, $('#progressBar'));
progressBar(80, $('#progressBar2'));
progressBar(60, $('#progressBar3'));

/* ==========================================================================
 #Mobile Menu
 ========================================================================== */

var $menu = $('#menu1'),
        $menulink = $('.menu-link');
$menulink.click(function () {
    $menulink.toggleClass('active');
    $menu.toggleClass('active');
    return false;
});

$('nav#menu1 a').click(function () {
    $('#menu1').removeClass('active');
});



/* ==========================================================================
 #iPad,iPhone,iPod Keyboard issue with position fixed
 ========================================================================== */
var iPad = navigator.userAgent.toLowerCase().indexOf("ipad");
var iPhone = navigator.userAgent.toLowerCase().indexOf("iphone");
var iPod = navigator.userAgent.toLowerCase().indexOf("ipod");
if (iPad > -1 || iPhone > -1 || iPod > -1)
{
    window.onscroll = function () {
        $('.totop-link').css('position', 'absolute');
        $('.totop-link').css('top', (window.pageYOffset + window.innerHeight - 39) + 'px');
    };
}





js/jquery.page.js
var pagetitle=$(document).find("title").text();

(function(window, undefined) {
			"use strict"; 	
	var Page	= (function() {
	
		var $container			= $( '#container' ),
			// the scroll container that wraps the articles
			$scroller			= $container.find( 'div.content-scroller' ),
			$menu				= $container.find( 'aside' ),
			// menu links
			$links				= $menu.find( 'nav > a' ),
			$articles			= $container.find( 'div.content-wrapper > article' ).not(".noscroll"),
			
		
			
			// button to scroll to the top of the page
			// only shown when screen size < 715
			$toTop				= $container.find( 'a.totop-link' ),
			// the browser nhistory object
			History 			= window.History,
			// animation options
			animation			= { speed : 800, easing : 'easeInOutExpo' },
			// jScrollPane options
			scrollOptions		= { verticalGutter : 0, hideFocus : true },
			// init function
			init				= function() {
				
				// initialize the jScrollPane on both the menu and articles
				_initCustomScroll();
				// initialize some events
				_initEvents();
				// sets some css properties 
				_layout();
				// jumps to the respective chapter
				// according to the url
				_goto();
				
			},
			_initCustomScroll	= function() {
			

				// Only add custom scroll to articles if screen size > 715.
				// If not the articles will be expanded
				if( $(window).width() > 767 ) {
				
					$articles.jScrollPane( scrollOptions );
				
				}
				// add custom scroll to menu
				$menu.children( 'nav' ).jScrollPane( scrollOptions );
			
			},
			_goto				= function( chapter ) {
				
				
				$(document).attr("title",pagetitle);
				
				
					// get the url from history state (e.g. chapter=3) and extract the chapter number
				var chapter 	= chapter || History.getState().url.queryStringToJSON().page,
					isHome		= ( chapter === undefined ),
					// we will jump to the introduction chapter if theres no chapter
					$article 	= $( chapter ? '#' + 'chapter' + chapter : '#' + 'introduction' );
				
				
				
					
					$('#link_introduction').removeClass('active');
					$('#link_research').removeClass('active');
					$('#link_publication').removeClass('active');
					$('#link_recognitions').removeClass('active');
					$('#link_services').removeClass('active');
					$('#link_teaching').removeClass('active');
					$('#link_team').removeClass('active');
					$('#link_positions').removeClass('active');
					$('#link_news').removeClass('active');
					
					$('#link_'+chapter).addClass('active');
					
				if( $article.length ) {
				
						// left / top of the element
					var left		= $article.position().left,
						top			= $article.position().top,
						// check if we are scrolling down or left
						// is_v will be true when the screen size < 715
						is_v		= ( $(document).height() - $(window).height() > 0 ),
						// animation parameters:
						// if vertically scrolling then the body will animate the scrollTop,
						// otherwise the scroller (div.content-scroller) will animate the scrollLeft
						param		= ( is_v ) ? { scrollTop : (isHome) ? top : top + $menu.outerHeight( true ) } : { scrollLeft : left },
						$elScroller	= ( is_v ) ? $( 'html, body' ) : $scroller;
					
					$elScroller.stop().animate( param, animation.speed, animation.easing, function() {
						
						// active class for selected chapter
						///$articles.removeClass( 'content-active' );
						//$article.addClass( 'content-active' );
						
						
					} );
				
				}
			
			},
			_saveState			= function( chapter ) {
			
				// adds a new state to the history object
				// this will trigger the statechange on the window
				if( History.getState().url.queryStringToJSON().page !== chapter ) {
						
					History.pushState( null, null, '?page=' + chapter );
		
					$('#link_introduction').removeClass('active');
					$('#link_about').removeClass('active');
					$('#link_skills').removeClass('active');
					$('#link_experience').removeClass('active');
					$('#link_education').removeClass('active');
					$('#link_portfolio').removeClass('active');
					$('#link_contact').removeClass('active');
					
					$('#link_'+chapter).addClass('active');
				
				}
			
			},
			_layout				= function() {
			
				// control the overflow property of the scroller (div.content-scroller)
				var windowWidth	= $(window).width();
				
				var isipad = navigator.userAgent.toLowerCase().indexOf("ipad");
				if(isipad > -1)
				{
					//$('.totop-link').hide();
					switch( true ) {
					case ( windowWidth <= 768 ) : $scroller.scrollLeft( 0 ).css( 'overflow', 'visible' ); break;
					case ( windowWidth <= 1024 ): $scroller.css( 'overflow-x', 'hidden' ); break;
					case ( windowWidth > 1024 ) : $scroller.css( 'overflow', 'hidden' ); break;
					};
				}
				else
				{
					switch( true ) {
					case ( windowWidth <= 768 ) : $scroller.scrollLeft( 0 ).css( 'overflow', 'visible' ); break;
					case ( windowWidth <= 1024 ): $scroller.css( 'overflow-x', 'hidden' ); break;
					case ( windowWidth > 1024 ) : $scroller.css( 'overflow', 'hidden' ); break;
					};
				}
			
			},
			_initEvents			= function() {
				
				_initWindowEvents();
				_initMenuEvents();
				_initArticleEvents();
				
			},
			_initWindowEvents	= function() {
			
				$(window).on({

					// when resizing the window we need to reinitialize or destroy the jScrollPanes
					// depending on the screen size
					'smartresize' : function( event ) {
						
						var isipad = navigator.userAgent.toLowerCase().indexOf("ipad");
						var ismobile =  navigator.userAgent.toLowerCase().indexOf("mobile");
						if(isipad > -1 || ismobile > -1 )
						{
						}
						else
						{
							_layout();
							
							$('article.content').not(".noscroll").each( function() {
								
								var $article 	= $(this),
									aJSP		= $article.data( 'jsp' );
								
								if( $(window).width() > 767 ) {
									
									( aJSP === undefined ) ? $article.jScrollPane( scrollOptions ) : aJSP.reinitialise();
									
									_initArticleEvents();
									
								}	
								else {
									
									// destroy article's custom scroll if screen size <= 715px
									if( aJSP !== undefined )
										aJSP.destroy();
									
									$container.off( 'click', 'article.content' );
									
								}
								
							});
							
							var nJSP = $menu.children( 'nav' ).data( 'jsp' );
							nJSP.reinitialise();
							
							// jumps to the current chapter
							_goto();
						}
					
					},
					// triggered when the history state changes - jumps to the respective chapter
					'statechange' : function( event ) {
						
						_goto();
					
					}
				});
			
			},
			_initMenuEvents		= function() {
			
				// when we click a menu link we check which chapter the link refers to,
				// and we save the state on the history obj.
				// the statechange of the window is then triggered and the page/scroller scrolls to the 
				// respective chapter's position
				$links.on( 'click', function( event ) {
					
					var href		= $(this).attr('href'),
						chapter		= ( href.search(/chapter/) !== -1 ) ? href.substring(8) : 0;
					
					_saveState( chapter );
					if (href.indexOf("#") != -1) 
						return false;
					else 
						return true;
					
						
				
				});
			
				// scrolls to the top of the page.
				// this button will only be visible for screen size < 715
				$toTop.on( 'click', function( event ) {
					
					$( 'html, body' ).stop().animate( { scrollTop : 0 }, animation.speed, animation.easing );
					
					return false;
				
				});
			
			},
			_initArticleEvents	= function() {
			
				// when we click on an article we check which chapter the article refers to,
				// and we save the state on the history obj.
				// the statechange of the window is then triggered and the page/scroller scrolls to the 
				// respective chapter's position
				
					
					
					if($(window).width()>768)
					{
						$container.on( 'click', 'article.content', function( event ) {
							
							var id			= $(this).attr('id'),
								chapter		= ( id.search(/chapter/) !== -1 ) ? id.substring(7) : 0;
							
							_saveState( chapter );
							
							//return false;
						
						});
					}
					
					
					
				
			
			};
			
		return { init : init };	
	
	})();
	
	Page.init();
	
})(window);

